docs/proposals/cluster-bootstrap-with-gossip.md

```diff
@@ -0,0 +1,153 @@
+
+If it doesn’t know an API server URL and cacert, it looks for a *bootstrap config* file (/etc/kubernetes/kubelet-bootstrap.json). If it finds one, it goes into **gossip** state.
+
+If it finds neither file, it goes into **pending** mode, where it waits for instructions from the kubeadm command.
+
+### Pending mode
+
+So, when the user (or the config management system of their choice) installs the OS packages, kubelet first starts and before any user has typed any kubeadm commands, it is in **pending** mode.
+
+The user then types kubeadm init master on the first server, the kubeadm binary writes a JSON representation of its arguments to /etc/kubernetes/kubelet-bootstrap.json.
+
+### Generic kubeadm functionality
+
+In the kubeadm command, in general:
+
+1. If --discovery flag is "gossip", which is the default:
+
+    1. If --token has been provided, remember it.
+
+    2. If --token has not been provided, generate a random string and print it to the user.
+
+    3. Write the following to /etc/kubernetes/kubelet-bootstrap.json:
+
+```
+{
+    "apiVersion": "v1alpha1",
+    "kind": "Discovery",
+    "role": "<role>",
+    "discovery": {
+        "type": "Gossip",
+        "token": "<token>",
+        "peers": []
+    }
+}
+```
+
+2. If --discovery flag is "out-of-band", require --apiserver-urls and --cacert arguments to also be passed.
+
+    4. In this case, disable the gossip server, and go straight into **bootstrapping** mode.
+
+    5. Write the following to /etc/kubernetes/kubelet-bootstrap.json:
+
+```
+{
+    "apiVersion": "v1alpha1",
+    "kind": "Discovery",
+    "role": "<role>",
+    "discovery": {
+        "type": "OutOfBand",
+        "apiserverURLs": ["https://foo/"],
+        "cacertFile": "/path/to/cacert.pem"
+    }
+}
+```
+
+In future, there could be other --discovery implementations, such as DNS- or discovery service-based.
+
+The Golang Discovery interface should be:
+
+```
+type Discovery interface {
+    Start()
+    Discover() (apiServerUrls []string, cacert x509.Certificate, err error)
+}
+```
+
+Start should be called every time the kubelet starts up, and it should return immediately, in case the discovery mechanism needs to run a persistent service (like a gossip socket).
+
+### Kubelet awakens
+
+The kubelet waits in **pending** mode for the bootstrap file to show up. When it does, it unmarshals the Discovery object from JSON depending on the "kind" field, and calls `Start()` and then `Discover()` on it.
+
+When it does, it's the Discovery implementation's job to move into **bootstrapping**.
+
+#### Masters
+
+Initializing a new master shouldn't require TLS bootstrap, because the master already has privileged access to the API server. New masters have to add their own address into the API server URL list in the mesh case.
```

Comment:
Yes, in this scheme the first master would generate the CA cert and put its public key out on the gossip network so that the joining nodes can trust and know they're talking to the real API server.

Advanced users may want to specify this cert, we shouldn't stop them. In fact, the "out of band" option will allow them to do exactly this.

We may also want to allow the user to provide a DNS name for the API server, and to have the initial cert have that added as subjectAltName etc. The default "I know nothing" assumption would be to put the IP address of the first master into gossip as the initial API server address.